src/main/java/com/basho/riak/client/core/codec/TermToBinaryCodec.java

```diff
@@ -0,0 +1,294 @@
+            throws OtpErlangDecodeException
+    {
+        final int colNameCount = is.read_list_head();
+        final String[] columnNames = new String[colNameCount];
+        for (int colNameIdx = 0; colNameIdx < colNameCount; colNameIdx++)
+        {
+            final String colName = new String(is.read_binary(), StandardCharsets.UTF_8);
+            columnNames[colNameIdx] = colName;
+
+            final boolean isLastRow = colNameIdx + 1 == colNameCount;
+            if (isLastRow)
+            {
+                is.read_nil();
+            }
+        }
+
+
+        final int colTypeCount = is.read_list_head();
+        assert (colNameCount == colTypeCount);
+        final String[] columnTypes = new String[colTypeCount];
+
+        for (int colTypeIdx = 0; colTypeIdx < colTypeCount; colTypeIdx++)
+        {
+            final String colType = is.read_atom();
+            columnTypes[colTypeIdx] = colType;
+
+            final boolean isLastRow = colTypeIdx + 1 == colNameCount;
+            if (isLastRow)
+            {
+                is.read_nil();
+            }
+        }
+
+        final ArrayList<RiakTsPB.TsColumnDescription> columnDescriptions = new ArrayList<>(colNameCount);
+        for (int colDescIdx = 0; colDescIdx < colNameCount; colDescIdx++)
+        {
+
+            final RiakTsPB.TsColumnDescription desc = RiakTsPB.TsColumnDescription.newBuilder().setName(
+                    ByteString.copyFromUtf8(columnNames[colDescIdx])).setType(RiakTsPB.TsColumnType.valueOf(
+                    columnTypes[colDescIdx].toUpperCase(Locale.US))).build();
+            columnDescriptions.add(desc);
+        }
+        return columnDescriptions;
+    }
+
+    private static ArrayList<RiakTsPB.TsRow> parseRows(OtpInputStream is, List<RiakTsPB.TsColumnDescription> columnDescriptions)
+            throws OtpErlangDecodeException, InvalidTermToBinaryException
+    {
+        final int rowCount = is.read_list_head();
+        final ArrayList<RiakTsPB.TsRow> rows = new ArrayList<>(rowCount);
+
+        for (int rowIdx = 0; rowIdx < rowCount; rowIdx++)
+        {
+            final boolean isLastRow = rowIdx + 1 == columnDescriptions.size();
+            rows.add(parseRow(is, columnDescriptions, isLastRow));
+        }
+        return rows;
+    }
+
+    private static RiakTsPB.TsRow parseRow(OtpInputStream is, List<RiakTsPB.TsColumnDescription> columnDescriptions, Boolean isLastRow)
+            throws OtpErlangDecodeException, InvalidTermToBinaryException
+    {
+        final int rowDataCount = is.read_tuple_head();
+        assert (columnDescriptions.size() == rowDataCount);
+
+        final Cell[] cells = new Cell[rowDataCount];
+        for (int j = 0; j < rowDataCount; j++)
+        {
+            final OtpErlangObject cell = is.read_any();
+            cells[j] = parseCell(columnDescriptions, j, cell);
+        }
+
+        if (isLastRow)
+        {
+            is.read_nil();
```

Comment:
This could be movef after the `for` loop in `parseRows`. The `isLastRow` parameter wouldn't be necessary then.